mapping_parenting_tech/analysis/prototyping/play_store/get_play_store_apps.py

```diff
@@ -7,141 +7,413 @@
+    for app_id in tqdm(
+        app_id_list,
+        desc="Retrieving app details",
+    ):
+        try:
+            app_details = app(
+                app_id, lang="en", country="gb"  # defaults to 'en'  # defaults to 'us'
+            )
+            all_app_details.update({app_id: app_details})
+
+        except Exception as e:  # needs modifying to capture specific errors
+            logging.info(f"Error on app id {app_id}: {e} {repr(e)}")
+            remove_apps.append(app_id)
+
+    for app_id in remove_apps:
+        all_app_details.remove(app_id)
+
+    return all_app_details
+
+
+# %%
+app_list = load_app_ids("outputs/data/top_free_parenting.pickle")
+
+# %%
+app_details = get_playstore_app_details(app_list)
 
-logging.info(
-    f"There are {review_counter} reviews available for {len(output_ids['Parenting apps'])} apps. {len(remove_apps)} apps were removed"
-)
-logging.info(
-    f"App id {most_reviews['id']} had the most reviews with {most_reviews['count']}"
-)
 
 # %% [markdown]
-# Save app details
+# ## Save and load app details to/from JSON file
+
+# %%
+def save_app_details(details_dict: dict, filename: str) -> bool:
+    """
+    Saves JSON representation of app details. Will append to file <filename> if it already exists.
+    Returns True if exits successfully
+    """
+
+    output_target = PROJECT_DIR / "outputs/data/" / filename
+    with open(output_target, "at") as output_file:  # append in text mode
+        json.dump(details_dict, output_file, indent=2, default=str)
+
+    return True
+
 
 # %%
-output_target = OUTPUT_PATH / OUTPUT_PLAY_DETAILS_FILE
-with open(output_target, "w") as output_file:
-    json.dump(all_app_details, output_file, indent=2, default=str)
-output_file.close
+def load_app_details(filename: str) -> json:
+    """
+    Loads JSON file containing app details from <filename>. Returns JSON object.
+    """
+
+    result = {}
+    with open(PROJECT_DIR / filename, "rt") as input_file:
+        result = json.load(input_file)
+
+    return result
 
-print(f"{len(all_app_details)} apps' details saved to {output_target}")
 
 # %% [markdown]
-# Download reviews for the apps in the list
+# ## Download reviews for the apps in a given list of app ids
+# These functions enable iterative downloading into a single file, which is appended as more reviews are added
+#
+# `get_playstore_app_reviews` gets up to 200 reviews for a single app and returns them as a list of dicts (JSON), while `get_playstore_app_list_reviews` retrievs reviews for a given list of apps and saves them to a CSV file.
 
 # %%
-all_app_reviews = dict()
-running_total = 0
-
-for app_id in tqdm(output_ids["Parenting apps"], desc="Retrieving app reviews"):
-    try:
-        app_reviews = reviews_all(
-            app_id,
-            sleep_milliseconds=0,  # defaults to 0
+def get_playstore_app_reviews(
+    target_app_id, how_many: int = 200, continuation_token: str = None
+) -> [[{}], object]:
```

Comment:
I think this would be more accurate:
```suggestion
) -> (Iterator[dict], object):
```

NB: Need to also add `from typing import Iterator` at the top of the notebook